Controller forced reconfiguration

joe-redpanda opened the pull request:
Adds the core logic for force reconfiguring the controller in the case of a cluster quorum loss.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
### Improvements

* Adds core logic for controller forced reconfiguration feature

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### 

* Short description of how this PR improves existing behavior.

-->

joe-redpanda committed:
consensus: fix force reconfig documentation

Forced reconfiguration does not append to the log on calling force
reconfigure. It is volatile

joe-redpanda committed:
cluster_test_fixture: add seed node to create node

Currently create_node_application points the newly created node to
node_id: 0 to join the cluster. This commit adds a seed_node_id to allow
this to be a different cluster node. Used in subsequent commits.

joe-redpanda committed:
controller: wire in CFRM

Wires in CFRM and adds top level api to invoke
controller_forced_reconfiguration.

joe-redpanda committed:
controller: add cfr smoke tests

Adds fixture tests which execute the process of controller forced
recovery.

steps
1. induce controller loss
2. reboot survivors to recovery mode
3. force recover the cluster to the survivors
4. add nodes until the cluster is back to original node number
5. reboot the restored cluster normally
6. check that leadership is restored on all data partitions

Adds cancellation test which validates that multiple invocations lead to
the old one being cancelled

joe-redpanda commented on src/v/cluster/controller.cc:
```diff
@@ -1305,4 +1309,213 @@ ss::future<> controller::clear_raft_manager_remake_cb() {
+        }
+        auto& quorum_loss_partitions = maybe_quorum_loss_partitions.value();
+
+        // todo find reasonable timeout
```
thoughts?

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--2f055245-840f-40ba-8e9f-9e15db20b57a-->
<!--TEST RESULTS-->
<!--9c2380dc-0db9-4420-b2b6-a5aed6a7139e-->
<!--TEST RESULTS-->
<!--e31d1307-91b3-48dc-978c-341e55cdb445-->
<!--TEST RESULTS-->
<!--88888c9b-30d8-47d1-a707-a2a068576da4-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73809">build#73809</a></summary>


| test_class                    | test_method            | test_arguments                                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                       |
|:------------------------------|:-----------------------|:--------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_replication_basic | {"shuffle_leadership": false, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73809#0199c5ee-9ffb-4249-9435-d2e6d9ec9c19 | FLAKY         | 18/21    | upstream reliability is '95.80686149936467'. current run reliability is '85.71428571428571'. drift is 10.09258 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic |
| ShadowLinkingReplicationTests | test_topic_delete      | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}}                              | integration | https://buildkite.com/redpanda/redpanda/builds/73809#0199c5ea-867d-4ace-a54b-a6fffa6eb099 | FAIL          | 0/1      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_topic_delete      |
| DataMigrationsApiTest         | test_mount_inexistent  | null                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/73809#0199c5ea-8683-4935-ad84-b4b5b807ed68 | FLAKY         | 19/21    | upstream reliability is '99.56989247311829'. current run reliability is '90.47619047619048'. drift is 9.0937 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_mount_inexistent          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74224">build#74224</a></summary>


| test_class           | test_method                                  | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                             |
|:---------------------|:---------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests | test_rapid_shadow_link_toggling              | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74224#0199e54c-fd53-49c3-8bf3-3fab5c1cb931 | FLAKY         | 17/21    | upstream reliability is '98.55595667870037'. current run reliability is '80.95238095238095'. drift is 17.60358 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling       |
| SegmentMsTest        | test_segment_rolling_with_retention_consumer | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74224#0199e54c-fd56-4bf2-8fd9-355926e5bbf9 | FLAKY         | 20/21    | upstream reliability is '91.50326797385621'. current run reliability is '95.23809523809523'. drift is -3.73483 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
| QuickTerminateTest   | test_terminate                               | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74224#0199e54c-fd5d-43c6-a029-1fc0876539fb | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=QuickTerminateTest&test_method=test_terminate                          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74340">build#74340</a></summary>


| test_class               | test_method                                      | test_arguments                                | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                            |
|:-------------------------|:-------------------------------------------------|:----------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LeaderEpochTest          | TestGetLeaderEpochWhileReconciling               |                                               | unit        | https://buildkite.com/redpanda/redpanda/builds/74340#0199eed5-dde9-44c2-a8c2-aee5a6e0f61e | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                                         |
| NodesDecommissioningTest | test_flipping_decommission_recommission          | {"cloud_topic": true, "node_is_alive": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74340#0199ef11-1627-4992-af6a-df3953dbb4bb | FLAKY         | 13/21    | upstream reliability is '81.0'. current run reliability is '61.904761904761905'. drift is 19.09524 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_flipping_decommission_recommission          |
| NodesDecommissioningTest | test_multiple_decommissions                      | {"cloud_topic": true}                         | integration | https://buildkite.com/redpanda/redpanda/builds/74340#0199ef11-1634-4637-b08a-1de2cb18ead3 | FLAKY         | 11/21    | upstream reliability is '63.51351351351351'. current run reliability is '52.38095238095239'. drift is 11.13256 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_multiple_decommissions                      |
| NodesDecommissioningTest | test_recommissioning_node                        | {"cloud_topic": true}                         | integration | https://buildkite.com/redpanda/redpanda/builds/74340#0199ef11-1632-45a7-915d-e595f72bf555 | FLAKY         | 14/21    | upstream reliability is '84.0'. current run reliability is '66.66666666666666'. drift is 17.33333 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recommissioning_node                        |
| NodesDecommissioningTest | test_recommissioning_one_of_decommissioned_nodes | {"cloud_topic": true}                         | integration | https://buildkite.com/redpanda/redpanda/builds/74340#0199ef11-1638-4b96-9060-3473fb52e034 | FLAKY         | 10/21    | upstream reliability is '91.25'. current run reliability is '47.61904761904761'. drift is 43.63095 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recommissioning_one_of_decommissioned_nodes |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74434">build#74434</a></summary>


| test_class                                     | test_method                                    | test_arguments                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                               |
|:-----------------------------------------------|:-----------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests                           | test_rapid_shadow_link_toggling                | null                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/74434#0199f3df-3172-4ce1-97cc-8f92f08ebd0d | FLAKY         | 1/21     | upstream reliability is '95.69536423841059'. current run reliability is '4.761904761904762'. drift is 90.93346 and the allowed drift is set to 50. The test should FAIL | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling         |
| ShadowLinkingReplicationTests                  | test_replication_with_failures                 | null                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/74434#0199f3df-3178-4616-8501-666382bfe8e9 | FLAKY         | 4/21     | upstream reliability is '78.1413612565445'. current run reliability is '19.047619047619047'. drift is 59.09374 and the allowed drift is set to 50. The test should FAIL | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures |
| EndToEndShadowIndexingTest                     | test_reset_from_cloud                          | {"cloud_storage_type": 2}                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74434#0199f3df-3171-492d-8eb9-254464da1456 | FAIL          | 0/1      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndShadowIndexingTest&test_method=test_reset_from_cloud             |
| EndToEndShadowIndexingTest                     | test_reset_from_cloud                          | {"cloud_storage_type": 1}                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74434#0199f3df-3172-4ce1-97cc-8f92f08ebd0d | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndShadowIndexingTest&test_method=test_reset_from_cloud             |
| TieredStorageTest                              | test_tiered_storage                            | {"cloud_storage_type_and_url_style": [2, "virtual_host"], "test_case": {"name": "(TS_Read == True, TS_Timequery == True)"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74434#0199f3df-3172-4ce1-97cc-8f92f08ebd0d | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TieredStorageTest&test_method=test_tiered_storage                        |
| WriteCachingFailureInjectionE2ETest            | test_crash_all                                 | {"use_transactions": false}                                                                                                 | integration | https://buildkite.com/redpanda/redpanda/builds/74434#0199f3a0-18fa-417d-8e47-886e857f6c97 | FLAKY         | 17/21    | upstream reliability is '82.11968276856526'. current run reliability is '80.95238095238095'. drift is 1.1673 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all           |
| src/v/cluster/tests/shard_placement_table_test | src/v/cluster/tests/shard_placement_table_test |                                                                                                                             | unit        | https://buildkite.com/redpanda/redpanda/builds/74434#0199f382-0750-4652-a039-1d07cdff7b00 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                            |
</details>
<!--/body-->

mmaslankaprv commented:
I really like this. The idea is simple and straight forward

bharathv commented on src/v/cluster/controller.h:
```diff
@@ -277,6 +277,9 @@ class controller {
 
     static bytes invariants_key();
 
+    ss::future<std::error_code> initialize_controller_forced_recovery(
```
nit: consider adding some high level notes on how this is supposed to work and how this is different from kafka partition forced reconfiguration.

bharathv commented on src/v/cluster/controller.cc:
```diff
@@ -1305,4 +1309,213 @@ ss::future<> controller::clear_raft_manager_remake_cb() {
+            vlog(
+              clusterlog.warn,
+              "controller forced recovery is a last resort, use the existing "
+              "controller leader: {} to repair the cluster",
```
nit: consider rephrasing so an operator can follow.. something like .. recovery only works if the controller leader has a loss of quorum.. use node wise recovery if partitions are leaderless..

bharathv commented on src/v/cluster/controller.cc:
```diff
@@ -1305,4 +1309,213 @@ ss::future<> controller::clear_raft_manager_remake_cb() {
+    }
+
+    { // step 1, forcibly reconfigure raft0
+        vlog(clusterlog.info, "starting raft0 force reconfiguration");
```
nit: include <dead_nodes> in the log output, also current configuration.

bharathv commented on src/v/cluster/controller.cc:
```diff
@@ -1305,4 +1309,213 @@ ss::future<> controller::clear_raft_manager_remake_cb() {
+          "recovery mode is a precondition for controller force recovery");
+        co_return cluster::errc::invalid_request;
+    }
+
```
do we need a mutex to ensure only one request is happening at any point?

joe-redpanda replied:
Good idea.

the behavior I would want seems to match the the single fiber executor. In lieu of waiting for that to merge, lets just do a mutex with an abort sorce

bharathv commented on src/v/cluster/controller.cc:
```diff
@@ -1305,4 +1309,213 @@ ss::future<> controller::clear_raft_manager_remake_cb() {
+                ++current_group_configuration.revision_id())};
+            const auto ec
+              = co_await this->_raft0->force_replace_configuration_locally(
+                std::move(surviving_voters),
```
check surviving_voters is not empty?

joe-redpanda replied:
probably a good idea

bharathv commented on src/v/cluster/controller.cc:
```diff
@@ -1305,4 +1309,213 @@ ss::future<> controller::clear_raft_manager_remake_cb() {
+
+            // case not leader, loop
+            if (!maybe_leader) {
+                co_await ss::sleep(std::chrono::seconds{5});
```
abortable sleep

bharathv commented on src/v/cluster/controller.cc:
```diff
@@ -1305,4 +1309,213 @@ ss::future<> controller::clear_raft_manager_remake_cb() {
+              clusterlog.error,
+              "failed to gather list of partitions with quorum loss with "
+              "error: {}",
+              maybe_quorum_loss_partitions.error());
```
missing return / retry

joe-redpanda replied:
Is this recoverable? afaik there can be no concurrent modifications at this time courtesy of recovery mode, so no errors here would be retriable?

bharathv replied:
ya I don't think this should fail in recovery mode

bharathv commented on src/v/cluster/controller.cc:
```diff
@@ -1305,4 +1309,213 @@ ss::future<> controller::clear_raft_manager_remake_cb() {
+    { // step 1, forcibly reconfigure raft0
+        vlog(clusterlog.info, "starting raft0 force reconfiguration");
+        const auto& current_group_configuration = this->_raft0->config();
+        const auto& current_node_configuration
```
After looking at the PR I had another thought. I think we can make this a bit coordinated than letting rpk fire the same command at all the brokers.

Since topics_frontend is available at this point, we let rpk send this command to any broker in the surviving nodes and it does an RPC to all the brokers in the surviving nodes to ensure these preconditions are met and the let it do the n-way RPC to force replicate the configuration.

joe-redpanda replied:
I'll think on this.

joe-redpanda replied:
Per offline discussion, the muxing to each surviving broker will go in rpk

bharathv commented on src/v/cluster/controller.cc:
```diff
@@ -1305,4 +1309,213 @@ ss::future<> controller::clear_raft_manager_remake_cb() {
+                      dead_node,
+                      ec);
+                    continue;
+                }
```
I didn't follow this fully,  doesn't decommission automatically replicate remove_node() once the quorum is back up and balancer runs?

joe-redpanda replied:
Let me check, I didn't think so

joe-redpanda replied:
not as far as I can tell

joe-redpanda replied:
Does not work without remove

bharathv commented on src/v/cluster/controller.cc:
```diff
@@ -1305,4 +1309,213 @@ ss::future<> controller::clear_raft_manager_remake_cb() {
+        vlog(clusterlog.info, "finished all partition force recovery");
+    }
+
+    { // step 4 decommission and remove all dead brokers
```
I think we should also document what happens if each step is aborted.. presumably if it is aborted _after_ replicating the configuration, user has to run node wise recovery commands manually?

joe-redpanda replied:
Lets doc and discuss

Operator steps:
```
1. reboot brokers into recovery mode
2. all brokers CFR
3. reboot brokers into normal mode
```

CFR steps stages:
```
0. validation
1. raft reconfigure
2. elect new leader
3. batch_force_reconfigure partitions
4. decommission & remove dead nodes
```

CFR failures
```
before 0 & after 0
-> the remaining nodes should be able to proceed so long as this wasn't the last voter
after 1.
-> same as 0, this is volatile until a leader emerges
after 2 -- at this point the controller will come back courtesy of the raft configuration batch replicated when leadership is won
-> follower dies:
    -> permanently dies: we will probably need another bulk force reconfiguration
    -> temporarily: afaik the retry loop on the leadership balancer will eventually successfully force reconfigure
-> leader dies / is changed:
    -> this cfr halts, but a new leader will come up
    -> Operator would have to manually resume with rpk
after 3
-> follower dies: same as 2
-> leader dies / is changed: same as 2
during & after 4:
-> follower dies: same as 2
-> leader dies / is changed: same as 2
```

Good:
The commands here are idempotent (decom 1 -> decom 1 should be fine), and repeated partition force reconfiguration just does the latest.

Bad:
Biggest pain is obviously the controller leader fragility.
Options:
1. replicate a batch that persists intent to force reconfigure
2. leave it as is (for now)
    - raft config is persisted as of a new leader
    - a new controller leader will emerge, an operator could repair the rest with rpk
    - could be made nicer with a second api that just orchestrates 3,4

bharathv reviewed (commented):
niceee...

joe-redpanda committed:
cluster: add cfr manager

Adds the functionality of controller forced reconfiguration to
controller_forced_reconfiguration_manager.

This class is a sidecar component which encapsulates the logic of
performing a cfr operation.

CFR will proceed in the following four steps:
0. validate the broker is in recovery mode, and there is no controller
   leader
1. force reconfiguration of the raft0 group to exclude dead_nodes
2. poll for a newly elected controller leader
3. force reconfigure all partitions with quorum loss off dead_nodes
4. decommission and remove all dead nodes

joe-redpanda commented on src/v/cluster/controller_forced_reconfiguration_manager.cc:
```diff
@@ -0,0 +1,366 @@
+            return erasing;
+        });
+
+        if (surviving_voters.size() == 0) {
```
I'll fix these if this needs another revision

joe-redpanda committed:
cluster/types: add error code with description

Adds error_info which is a union of error_code and sstring. This will be
used to pass descriptive errors to an operator of CFR

joe-redpanda commented:
Changes:
1. cancellation_token and cancellation error code dropped
2. CFRM will now return error code on multiple CFR requests instead of cancelling the old one
3. CFR return swapped to error_info, which contains error code and a message for the operator
4. the force applied controller group will now be 2N-1 where N is the number of surviving brokers. This guarantees that all surviving brokers will be consulted before electing a new leader

bharathv commented on src/v/cluster/types.h:
```diff
@@ -3687,6 +3687,13 @@ struct get_current_cluster_epoch_response
 
+struct error_info {
+    std::error_code err;
+    ss::sstring message;
```
is this different from https://en.cppreference.com/w/cpp/error/error_code/message.html?

joe-redpanda replied:
yeah error message is a compile time map from error_code, error_domain -> message

this is dynamic because I'm shoving in information on what specifically went wrong such as which brokers failed to decom

bharathv commented on src/v/cluster/controller_forced_reconfiguration_manager.cc:
```diff
@@ -0,0 +1,360 @@
+
+    current_voters.resize(surviving_node_size * 2 - 1);
+
+    vlog(
```
is this missing some logic?

joe-redpanda replied:
No, but you are right that this logline is misplaced, I'll move this to right before the forced reconfiguration

bharathv commented on src/v/cluster/controller_forced_reconfiguration_manager.cc:
```diff
@@ -0,0 +1,360 @@
+                      dead_node,
+                      ec);
+                    continue;
+                }
```
I lost the original comment, can you explain why this is needed, I still can't convince myself this is needed. Just want to make sure I understand it.

joe-redpanda replied:
Not 100% sure but the test fails without on timeouts.

IIRC we discussed that the intent to perform a decommission is completed by the observation of a corresponding delete, so afaik this should just immediately complete the decommission.

joe-redpanda replied:
This ended up being an assumption made by wait_for_nodes

It assumes that the metadata number should be the same as the cluster instance count (true most of the time)

Here, decommission will not remove metadata until all replicas can be moved off of the dying nodes, which only completes when the cluster reboots into non-recovery mode.

Timeout was occurring because the check was evaluating expected: 3 against found: 5

resolution is a wait condition that is permissive to additional nodes in metadata.

Added a check that this is rectified by test end, though

bharathv commented on src/v/cluster/tests/controller_forced_reconfiguration_test.cc:
```diff
@@ -0,0 +1,424 @@
+          true,
+          false,
+          false,
+          /* cluster_linking_enabled */ true,
```
nit: remove

bharathv commented on src/v/cluster/tests/controller_forced_reconfiguration_test.cc:
```diff
@@ -0,0 +1,424 @@
+                  model::kafka_namespace,
+                  topic_config.name,
+                  model::partition_id{partition_number}};
+                vlog(_logger.info, "polling for leadership on ntp: {}", ntp);
+                auto [rp_thread, partition] = this->get_leader(ntp);
+                return rp_thread != nullptr && partition != nullptr;
+            });
```
wonder if this should also check all the the replicas are from the surviving et.

bharathv reviewed (commented):
some clarifying comments, generally looks fine to me.

joe-redpanda commented on src/v/cluster/controller_forced_reconfiguration_manager.cc:
```diff
@@ -0,0 +1,360 @@
+          return first.id() < second.id();
+      });
+
+    current_voters.resize(surviving_node_size * 2 - 1);
```
anchror

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74423-->
<!--8b64bbd6-7aa3-463d-9a5f-95460a89a92f-->
<!--f0516380-7827-4e36-92f1-0c2e21fbb2ce-->
<!--title-->
## Retry command for [Build#74423](https://buildkite.com/redpanda/redpanda/builds/74423)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTest.test_reset_from_cloud@{"cloud_storage_type":2}
tests/rptest/tests/archive_retention_test.py::CloudArchiveRetentionTest.test_delete@{"cloud_storage_type":2,"retention_type":"retention.ms"}
```
<!--/body-->

bharathv reviewed (approved):
great work 👍

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74434-->
<!--0330b5fc-cd84-4e56-903a-9eff53c81649-->
<!--e056c8d0-b08b-4a44-8d44-398a8c1e365d-->
<!--2a90bd1e-470c-412f-b8ae-9437a53d3f2e-->
<!--8368337e-ae48-4de6-a4fa-e57cd65e508a-->
<!--b43f7029-860b-4a9c-bc2c-b325e2457b99-->
<!--d4d0bbc4-1abb-4ced-a779-160eb1959398-->
<!--title-->
## Retry command for [Build#74434](https://buildkite.com/redpanda/redpanda/builds/74434)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTest.test_reset_from_cloud@{"cloud_storage_type":2}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkBasicTests.test_rapid_shadow_link_toggling
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTest.test_reset_from_cloud@{"cloud_storage_type":1}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[2,"virtual_host"],"test_case":{"name":"(TS_Read == True, TS_Timequery == True)"}}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_replication_with_failures
```
<!--/body-->

joe-redpanda commented:
Failed on unrelated functionality. Will kick of retry
